src/data/list/basic.lean

```diff
@@ -39,6 +39,10 @@ theorem tail_eq_of_cons_eq {h₁ h₂ : α} {t₁ t₂ : list α} :
       (h₁::t₁) = (h₂::t₂) → t₁ = t₂ :=
 assume Peq, list.no_confusion Peq (assume Pheq Pteq, Pteq)
 
+theorem cons_ne_self {α : Type u} : ∀ (a : α) (l : list α), a::l ≠ l
+| a nil := cons_ne_nil a nil
+| a (h :: t) := cons_ne_self h t ∘ tail_eq_of_cons_eq
```

Comment:
Alternative proof from zulip:
```suggestion
lemma cons_ne_self {α} (a : α) (l : list α) : list.cons a l ≠ l :=
mt (congr_arg list.length) (nat.succ_ne_self _)
```
I'm not super attached to which proof is used, I suppose it ultimately doesn't really matter

Reply:
+1 for Mario's proof.

A few places need to be changed for consistency: theorem, Type u, a :: l.